### PR TITLE
Encode link_id in tagged linkage

### LINKED ISSUE #48218: pkgimages: storage of `link_ids` should be compressed
Towards the end of writing out pkgimages, we write a section of `link_ids`: https://github.com/JuliaLang/julia/blob/1e5fdb29f8858f3244f6aff116ee12e4c8247f3a/src/staticdata.c#L2477-L2484

These store the buildid of the module for an external linkage; they act as an invariant name for a module we depend on, and are used for pointer relocation when we deserialize.

It turns out these are quite repetitious, and the large majority of such linkages are against the sysimage with buildid 0. Consequently we should be able to store them with some form of compression. This will only modestly shrink the size of pkgimages (maybe 5%ish), but it seems worth having.

### PR DESCRIPTION
**Pull request by vtjnash**
On 64-bit, we have enough space to encode (1) the tag, (2) the `depmods` index, and (3) the offset all in a single 64-bit pointer field. This means we don't need the external `link_id` arrays, which reduces the size of many pkgimages by ~5%.

On 32-bit, we don't have enough bits to implement this strategy. However, most linkages seem to be against the sysimage, and so by giving that a separate tag we can achieve similar compression because the `link_id` lists will be much shorter.

Fixes: https://github.com/JuliaLang/julia/issues/48218
Closes: https://github.com/JuliaLang/julia/pull/48279

### CODE REVIEW

**Comment by timholy** [comment]
I never did get around to finishing this. Am I understanding you're OK with the uglification compared to #48279? If so I can focus on this branch.

**Comment by vtjnash** [comment]
I didn't realize it wasn't finished before. It is fixed now. I think you will see the new diff is smaller and more complete than the old diff (and I even combined 2 existing tags, so we are net neutral) 🙂

File: src/codegen.cpp [hunk 1]
```diff
@@ -4251,7 +4251,7 @@ static jl_cgval_t emit_invoke(jl_codectx_t &ctx, const jl_cgval_t &lival, const
                     bool cache_valid = ctx.use_cache;
                     bool external = false;
                     if (ctx.external_linkage) {
-                       if (jl_object_in_image((jl_value_t*)codeinst)) {
+                       if (0 && jl_object_in_image((jl_value_t*)codeinst)) {
```
Comment (vchuravy) [inline src/codegen.cpp:4254]: Did you intend to commit this part?

Comment (KristofferC) [reply src/codegen.cpp:4254]: Add a comment or something then? Looks very weird to have that standing freely.

Comment (vchuravy) [reply src/codegen.cpp:4254]: Without this code we won't link from package images to the system image. Instead each package image will contain a copy of the native code, so I don't think disabling is an option.

Comment (fingolfin) [reply src/codegen.cpp:4254]: A comment here and in the other places doing this would indeed have been nice...

**Comment by KristofferC** [comment]
A s all size reduction does not seem necessary to backport. And this PR introduces a bunch of dead code so it feels like it needs various follow ups.

**Comment by vtjnash** [comment]
What deadcode is introduced here? I accidentally defined some pointers that used to be null, requiring a more explicit code guard, but that didn't change anything observable (indeed, the point of it was to avoid changing anything notable)

**Comment by KristofferC** [comment]
You know better than me but it does look weird to have a bunch of `if(0 && ...)` spread out around the code base...

**Pull request merged**